MES-6279: comms screen

andrewsetterfield opened the pull request:
## Description

MES-6279: Port comms page to DES4

## Checklist

- [ ] PR title includes the JIRA ticket number
- [ ] Branch is rebased against the latest develop
- [ ] Code has been tested manually
- [ ] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

## Screenshots (optional)
![Screenshot 2021-05-10 at 14 46 54](https://user-images.githubusercontent.com/48550267/117669401-c3089200-b19e-11eb-9a09-eabe43e7d6c0.png)
![Screenshot 2021-05-10 at 14 46 59](https://user-images.githubusercontent.com/48550267/117669405-c3a12880-b19e-11eb-9cee-10d27713f588.png)
![Screenshot 2021-05-10 at 14 47 06](https://user-images.githubusercontent.com/48550267/117669408-c439bf00-b19e-11eb-9e46-431ed74e423c.png)

updated
![image](https://user-images.githubusercontent.com/48550267/117781301-65288880-b238-11eb-81d7-d4dfcb4aa8ec.png)

andrewsetterfield committed:
add practice test base page

andrewsetterfield committed:
WIP

andrewsetterfield committed:
added category generic communications page

andrewsetterfield committed:
added unit tests for child components

andrewsetterfield committed:
added unit tests for child components

andrewsetterfield committed:
added unit tests for child components

andrewsetterfield committed:
completed unit tests

andrewsetterfield committed:
resolve conflict

tomcrawleyy commented on src/app/pages/communication/components/provided-email/provided-email.ts:
```diff
@@ -0,0 +1,44 @@
+      this.radioButtonControl = new FormControl('', Validators.required);
+      this.formGroup.addControl(ProvidedEmailComponent.radioCtrl, this.radioButtonControl);
+    }
+    this.radioButtonControl.patchValue(!!this.isProvidedEmailAddressChosen);
```
whats up with this being a double negative (!!)?

andrewsetterfield replied:
is not a double negative is a special kind of negative that includes undefined/null etc.

matthew2564 commented on src/app/pages/communication/communication.analytics.effects.ts:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  @Effect()
+  communicationViewDidEnter$ = this.actions$.pipe(
```
update to using new createEffect syntax

andrewsetterfield replied:
will do (thought I'd found them all)

matthew2564 commented on src/app/pages/communication/communication.page.scss:
```diff
@@ -0,0 +1,51 @@
+    background-color: var(--gds-lavendar);
+    font-weight: bold;
+    pointer-events: none;
+    color: var(--gds-black) !important;
```
are we supposed to be using !important?

andrewsetterfield replied:
can remove. have added host since adding that which fixes issue

matthew2564 commented on src/app/pages/communication/communication.page.ts:
```diff
@@ -1,30 +1,354 @@
-    private navController: NavController,
+    public platform: Platform,
+    public router: Router,
+    public authenticationProvider: AuthenticationProvider,
```
platform, router and authenticationProvider can have their access modifiers removed if just being used to inject into basepage

andrewsetterfield replied:
@matthew2564 - will remove. do u wanna add the same comment to mes-6277 review

matthew2564 replied:
@andrewsetterfield The same comment was added to MES-6277

matthew2564 commented on src/app/pages/communication/communication.page.ts:
```diff
@@ -1,30 +1,354 @@
+        select(getCandidateEmailAddress),
+        take(1),
+      ),
+      communicationEmail$: this.store$.pipe(
```
can we pipe from currentTest rather than from store here?

andrewsetterfield replied:
was taken directly from des3, but can take a look

andrewsetterfield replied:
updated to use current lest as per the rest of the file

matthew2564 commented on src/app/pages/communication/communication.page.ts:
```diff
@@ -1,30 +1,354 @@
+
+    this.subscription = this.merged$.subscribe();
+
+    this.initialiseDefaultSelections();
```
in DES3 this is encapsulated within `if (this.shouldPreselectADefaultValue()) {`

andrewsetterfield replied:
yes, was redundant

matthew2564 commented on src/app/pages/communication/components/new-email/__tests__/new-email.spec.ts:
```diff
@@ -0,0 +1,85 @@
+
+  describe('DOM', () => {
+
+    describe('i18n', () => {
```
tests appear different to DES3

andrewsetterfield replied:
yes. the formatting in DES3 was awful. tests wernt linked to functions, there were tests that covered functionility covered in its components.

matthew2564 commented on src/app/pages/communication/components/new-email/new-email.ts:
```diff
@@ -0,0 +1,70 @@
+  static readonly newEmail: string = 'newEmail';
+  static readonly newEmailCtrl: string = 'newEmailCtrl';
+  static readonly radioCtrl: string = 'radioCtrl';
+  maximumCallStackHandler = { emitEvent: false, onlySelf: true };
```
is this used or in DES3?

andrewsetterfield replied:
was using it by your suggestion :D after looking at search.  turned out not to be nessacery.

matthew2564 commented on tsconfig.json:
```diff
@@ -24,7 +24,9 @@
-      "@components/*": ["src/components/*"]
+      "@components/*": ["src/components/*"],
+      "@assets/*": ["src/assets/*"],
+      "@app-module/*": ["src/app/*"],
```
any reason for not just @app?

andrewsetterfield replied:
changed.

matthew2564 commented on src/app/pages/communication/components/privacy-notice/__tests__/privacy-notice.spec.ts:
```diff
@@ -0,0 +1,40 @@
+      ],
+      providers: [
+        { provide: TranslateService, useValue: translateServiceMock },
+        { provide: Config, useFactory: () => ConfigMock.instance() },
```
Don't think we need this Config declaration

andrewsetterfield replied:
funny thing is, i picked matt up for that yesterday. removed

matthew2564 commented:
Not sure we need the doc blocks inside src/app/pages/communication/communication.page.ts, seems needless considering rest of app omits them.

Also, the background of the page is still grey and not white as per Miro/DES3

andrewsetterfield committed:
addressed PR comments

andrewsetterfield commented:
@matthew2564  - agreed, the doc blocks dont add much value as they're not exactly informative. removed. Fixed background grey/white which highlighted a small margin issue which i fixed too.

andrewsetterfield committed:
rebase

andrewsetterfield committed:
updated selector to use current test, not store directly
